snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -94,7 +95,7 @@ def process_message(
                 "tags.key": _array_literal(keys),
                 "tags.value": _array_literal(values),
                 **self._process_values(message),
-                "retention_days": _literal(message["retention_days"]),
+                "retention_days": _literal(enforce_retention(message, timestamp)),
```

Comment:
Using the `enforce_retention` API from events_format, we introduce the following two constraints in sessions and metrics.
1. If a project has lower retention defined for either events or transactions, the lower retention would be applicable for sessions and metrics.
2. If the timestamp on the message is older than the retention days, we would crash the consumer since the `EventTooOld` exception would be raised.

I am not too familiar with the product to know if we want to add these additional constraints or not.

Reply:
@nikhars those are good points. 

> 1. If a project has lower retention defined for either events or transactions, the lower retention would be applicable for sessions and metrics.


So can a project have its own `LOWER_RETENTION_DAYS` set? or do you mean the `RETENTION_OVERRIDES`? I think if we have `ENFORCE_RETENTION` set to true, regardless of project, the retention days will be either `30` or `90` right? Or are you specifically pointing out that we may want to have different retention days defined for different datasets? Like have events/transactions lower retention be 30 while metrics/sessions be 90 or something like that

> 2. If the timestamp on the message is older than the retention days, we would crash the consumer since the EventTooOld exception would be raised.

Yeah, it seems like from the metrics/sessions queries in sentry we hardcode in a time delta of 90days, so I was assuming the product isn't querying for that data anyway so we shouldn't be processing it. However, I see in the events_processor_base we handle that exception, so I should probs add handling it in the metrics and sessions processors as well

Reply:
> So can a project have its own LOWER_RETENTION_DAYS set? or do you mean the RETENTION_OVERRIDES? I think if we have ENFORCE_RETENTION set to true, regardless of project, the retention days will be either 30 or 90 right? Or are you specifically pointing out that we may want to have different retention days defined for different datasets? Like have events/transactions lower retention be 30 while metrics/sessions are 90 or something like that

Each project can set a lower retention period by using RETENTION_OVERRIDES. The RETENTION_OVERRIDES is currently set at 30 for certain projects. RETENTION_OVERRIDES were only being used for errors/transactions. So the question is if a project has lower retention for errors/transactions via RETENTION_OVERRIDES, should it also have lower retention on sessions/metrics?

Reply:
> So the question is if a project has lower retention for errors/transactions via RETENTION_OVERRIDES, should it also have lower retention on sessions/metrics?

No. Retention have to be independent.